packaging/python3-defaults/debian/py3versions.1
.TH PY3VERSIONS
.SH NAME
py3versions \- print python3 version information
.SH SYNOPSIS
.PP
.B py3versions
.I [-h] [-v] [<options>]
.SH DESCRIPTION
.PP
.B py3versions
prints information about installed, supported python3 runtimes, the
default runtime, and parses the information of the PythonVersion
fields in the package control file for Python3.
.SH OPTIONS
.TP
.I -d, --default
Show the default python3 version.
.TP
.I -s, --supported
Show the supported python3 versions.
.TP
.I -r, --requested [<version string>|<control file>]
Reads the value of the
.B X-Python3-Version
field in the source section of a control file and shows all matching
python3 versions. The parameter is interpreted as a version string, if
it is not the name of a file. If the X-Python3-Version field is
missing, it will fall back to list the supported Python3
versions after checking debian/control. Versions less than 3 are
ignored. The keyword all is ignored because it covers all Python2
versions.  The former keyword current is not supported at all in
Python3.
.TP
.I -i, --installed
Show the installed supported python3 versions.
.TP
.I -v, --version
Limit the output to the version numbers of the python3 versions.
.TP
.I -h, --help
Print a help text.
.SH SEE ALSO
Python policy.
.SH AUTHOR
Matthias Klose


vim/NEWS
vim (2:7.3.154+hg~74503f6ee649-1) unstable; urgency=low

  The vim-lesstif package has been removed in favor of the new vim-athena
  package.  The intent behind both packages is to provide a lighter-weight GUI
  package as well as one that allows using XFLD fonts.  The Athena toolkit,
  however,  has broader usage and reduces divergences with downstream
  distributions.

 -- James Vega <jamessan@debian.org>  Sun, 27 Feb 2011 12:45:40 -0500

vim (2:7.2c.000-1) experimental; urgency=low

  The autoindent option is no longer enabled by default because:
  - 'autoindent' is rather naïve and turning on filetype-specific indentation
    (via "filetype indent on") is usually a better choice.
  - not having any automatic indenting enabled by default prevents new users
    from encountering the bad indentation issue when pasting text.

  Filetype plugins are no longer enabled by default because certain actions
  (like preventing the loading of the default menus) *must* occur before
  filetype detection is enabled.  By enabling filetype plugins (and detection)
  in debian.vim, we are preventing the user from being able to make changes to
  Vim's behavior.  Filetype plugins can be re-enabled via
  "filetype plugin on".

 -- James Vega <jamessan@debian.org>  Thu, 07 Aug 2008 12:25:39 -0400

vim (1:7.1.285-1) unstable; urgency=low

  Due to popular demand (and general conformity with other filetype plugins),
  the debchangelog and debcontrol filetype plugins no longer enable folding by
  default.  If you wish to enable their folding, you can add the following
  commands to your ~/.vimrc for debchangelog and debcontrol, respectively:

    let g:debchangelog_fold_enable = 1
    let g:debcontrol_fold_enable = 1

 -- James Vega <jamessan@debian.org>  Sun, 30 Mar 2008 12:42:35 -0400

vim (1:7.1-022+1) unstable; urgency=low

  * /usr/share/vim/addons/ is no longer in the vim runtimepath

    This is intended to avoid arbitrarly clashes among addons installed
    system-wide and enabled by default, and to set up a common ground for Vim
    extensions of various kinds.  From now on vim addons won't be enabled by
    default upon installation but will need to be enabled on a per-addon basis
    either in a system-wide manner (affecting all users of the system) or by
    each user who wants to use them.  A tool exists for helping with addon
    management: it is called "vim-addons" and is shipped by the package
    "vim-addon-manager"; all vim addon packages should recommend it.

    For more information see the VIM packaging policy (available on the web at
    http://pkg-vim.alioth.debian.org/vim-policy.html/index.html) and the
    manual page of vim-addons, available in the vim-addon-manager package.

 -- Stefano Zacchiroli <zack@debian.org>  Thu, 19 Jul 2007 23:28:44 +0200


packaging/python3-defaults/debian/py3versions.py
#! /usr/bin/python3

import os
import re
import sys

_defaults = None
_old_versions = None
_unsupported_versions = None
_supported_versions = ["python%s" % ver.strip() for ver in
                       os.environ.get('DEBPYTHON3_SUPPORTED', '').split(',')
                       if ver.strip()]
#_default_version = "python%s" % os.environ.get('DEBPYTHON3_DEFAULT', '')
#if _default_version == 'python':
#    _default_version = None
_default_version = None


def read_default(name=None):
    global _defaults
    from configparser import ConfigParser, NoOptionError
    if not _defaults:
        if os.path.exists('/usr/share/python3/debian_defaults'):
            config = ConfigParser()
            defaultsfile = open('/usr/share/python3/debian_defaults')
            config.read_file(defaultsfile)
            defaultsfile.close()
            _defaults = config

    if _defaults and name:
        try:
            value = _defaults.get('DEFAULT', name)
        except NoOptionError:
            raise ValueError
        return value
    return None


def parse_versions(vstring):
    if len(vstring.split(',')) > 2:
        raise ValueError('too many arguments provided for X-Python3-Version: min and max only.')
    import operator
    operators = {None: operator.eq, '=': operator.eq,
                 '>=': operator.ge, '<=': operator.le,
                 '<<': operator.lt}
    vinfo = {}
    exact_versions = set()
    version_range = set(supported_versions(version_only=True))
    relop_seen = False
    for field in vstring.split(','):
        field = field.strip()
        if field == 'all':
            continue
        if field in ('current', 'current_ext'):
            continue
        vinfo.setdefault('versions', set())
        ve = re.compile('(>=|<=|<<|=)? *(\d\.\d)$')
        m = ve.match(field)
        try:
            if not m:
                raise ValueError('error parsing Python3-Version attribute')
            op, v = m.group(1), m.group(2)
            vmaj, vmin = v.split('.')
            if int(vmaj) < 3:
                continue
            if op in (None, '='):
                exact_versions.add(v)
            else:
                relop_seen = True
                filtop = operators[op]
                version_range = [av for av in version_range if filtop(av, v)]
        except Exception:
            raise ValueError('error parsing Python3-Version attribute')
    if 'versions' in vinfo:
        vinfo['versions'] = exact_versions
        if relop_seen:
            vinfo['versions'] = exact_versions.union(version_range)
    return vinfo


def old_versions(version_only=False):
    global _old_versions
    if not _old_versions:
        try:
            value = read_default('old-versions')
            _old_versions = [s.strip() for s in value.split(',')]
        except ValueError:
            _old_versions = []
    if version_only:
        return [v[6:] for v in _old_versions]
    else:
        return _old_versions


def unsupported_versions(version_only=False):
    global _unsupported_versions
    if not _unsupported_versions:
        try:
            value = read_default('unsupported-versions')
            _unsupported_versions = [s.strip() for s in value.split(',')]
        except ValueError:
            _unsupported_versions = []
    if version_only:
        return [v[6:] for v in _unsupported_versions]
    else:
        return _unsupported_versions


def supported_versions(version_only=False):
    global _supported_versions
    if not _supported_versions:
        try:
            value = read_default('supported-versions')
            _supported_versions = [s.strip() for s in value.split(',')]
        except ValueError:
            cmd = ['/usr/bin/apt-cache', '--no-all-versions',
                   'show', 'python3-all']
            try:
                import subprocess
                p = subprocess.Popen(cmd, bufsize=1,
                                     shell=False, stdout=subprocess.PIPE)
                fd = p.stdout
            except ImportError:
                fd = os.popen(' '.join(cmd))
            depends = None
            for line in fd:
                if line.startswith('Depends:'):
                    depends = line.split(':', 1)[1].strip().split(',')
            fd.close()
            depends = [re.sub(r'\s*(\S+)[ (]?.*', r'\1', s) for s in depends]
            _supported_versions = depends
    if version_only:
        return [v[6:] for v in _supported_versions]
    else:
        return _supported_versions


def default_version(version_only=False):
    global _default_version
    if not _default_version:
        _default_version = os.readlink('/usr/bin/python3')
    # consistency check
    debian_default = read_default('default-version')
    if not _default_version in (debian_default, os.path.join('/usr/bin', debian_default)):
        raise ValueError("the symlink /usr/bin/python3 does not point to the "
                         "python3 default version. It must be reset "
                         "to point to %s" % debian_default)
    _default_version = debian_default
    if version_only:
        return _default_version[6:]
    else:
        return _default_version


def requested_versions(vstring, version_only=False):
    versions = None
    vinfo = parse_versions(vstring)
    supported = supported_versions(version_only=True)
    if len(vinfo) == 1:
        versions = vinfo['versions'].intersection(supported)
    else:
        raise ValueError('No python3 versions in version string')
    if not versions:
        raise ValueError('empty set of versions')
    if version_only:
        return versions
    else:
        return ['python%s' % v for v in versions]


def installed_versions(version_only=False):
    import glob
    supported = supported_versions()
    versions = [os.path.basename(s)
                for s in glob.glob('/usr/bin/python3.[0-9]')
                if os.path.basename(s) in supported]
    versions.sort()
    if version_only:
        return [v[6:] for v in versions]
    else:
        return versions


class ControlFileValueError(ValueError):
    pass


class MissingVersionValueError(ValueError):
    pass


def extract_pyversion_attribute(fn, pkg):
    """read the debian/control file, extract the X-Python3-Version
    field."""

    version = None
    sversion = None
    section = None
    with open(fn, encoding='utf-8') as controlfile:
        lines = [line.strip() for line in controlfile]
    for line in lines:
        if line == '':
            if pkg == 'Source':
                break
            section = None
        elif line.startswith('Source:'):
            section = 'Source'
        elif line.startswith('Package: ' + pkg):
            section = pkg
        elif line.lower().startswith('x-python3-version:'):
            if section != 'Source':
                raise ValueError('attribute X-Python3-Version not in Source section')
            sversion = line.split(':', 1)[1].strip()
    if section is None:
        raise ControlFileValueError('not a control file')
    if pkg == 'Source':
        if sversion is None:
            raise MissingVersionValueError('missing X-Python3-Version in control file')
        return sversion
    return version


'''
def requested_versions_bis(vstring, version_only=False):
    versions = []
    py_supported_short = supported_versions(version_only=True)
    for item in vstring.split(','):
        v=item.split('-')
        if len(v)>1:
            if not v[0]:
                v[0] = py_supported_short[0]
            if not v[1]:
                v[1] = py_supported_short[-1]
            for ver in py_supported_short:
                try:
                    if version_cmp(ver,v[0]) >= 0 \
                           and version_cmp(ver,v[1]) <= 0:
                        versions.append(ver)
                except ValueError:
                    pass
        else:
            if v[0] in py_supported_short:
                versions.append(v[0])
    versions.sort(version_cmp)
    if not versions:
        raise ValueError('empty set of versions')
    if not version_only:
        versions=['python'+i for i in versions]
    return versions
'''


def main():
    from optparse import OptionParser
    usage = '[-v] [-h] [-d|--default] [-s|--supported] [-i|--installed] '
    '[-r|--requested <version string>|<control file>]'
    parser = OptionParser(usage=usage)
    parser.add_option('-d', '--default',
                      help='print the default python3 version',
                      action='store_true', dest='default')
    parser.add_option('-s', '--supported',
                      help='print the supported python3 versions',
                      action='store_true', dest='supported')
    parser.add_option('-r', '--requested',
                      help='print the python3 versions requested by a build; '
                           'the argument is either the name of a control file '
                           'or the value of the X-Python3-Version attribute',
                      action='store_true', dest='requested')
    parser.add_option('-i', '--installed',
                      help='print the installed supported python3 versions',
                      action='store_true', dest='installed')
    parser.add_option('-v', '--version',
                      help='print just the version number(s)',
                      default=False, action='store_true', dest='version_only')
    opts, args = parser.parse_args()
    program = os.path.basename(sys.argv[0])

    if opts.default and len(args) == 0:
        try:
            print(default_version(opts.version_only))
        except ValueError as msg:
            print("%s:" % program, msg)
            sys.exit(1)
    elif opts.supported and len(args) == 0:
        print(' '.join(supported_versions(opts.version_only)))
    elif opts.installed and len(args) == 0:
        print(' '.join(installed_versions(opts.version_only)))
    elif opts.requested and len(args) <= 1:
        if len(args) == 0:
            versions = 'debian/control'
        else:
            versions = args[0]
        try:
            if os.path.isfile(versions):
                fn = versions
                try:
                    vstring = extract_pyversion_attribute(fn, 'Source')
                    vs = requested_versions(vstring, opts.version_only)
                except ControlFileValueError:
                    sys.stderr.write("%s: not a control file: %s, "
                                     % (program, fn))
                    sys.exit(1)
                except MissingVersionValueError:
                    sys.stderr.write("%s: no X-Python3-Version in control "
                                     "file, using supported versions\n" %
                                     program)
                    vs = supported_versions(opts.version_only)
            else:
                vs = requested_versions(versions, opts.version_only)
            print(' '.join(vs))
        except ValueError as msg:
            sys.stderr.write("%s: %s\n" % (program, msg))
            sys.exit(1)
    else:
        sys.stderr.write("usage: %s %s\n" % (program, usage))
        sys.exit(1)

if __name__ == '__main__':
    main()
